docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,101 @@
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+block_configs:
+  - block_type: LLMBlock
+    block_config: # LLMBlock constructor kwargs
+      block_name: gen_knowledge
+      output_cols: ["question", "response"]
+    gen_kwargs: # kwargs for block.generate()
+      max_tokens": 2048,
+    drop_duplicates: ["question"]
+  - block_type: FilterByValueBlock
+    block_config:
+      block_name: filter_faithfulness
+      filter_column": judgment
+      filter_value: YES
+      operation: operator.eq
+      batch_kwargs:
+        num_procs: 8
+     drop_columns: ["judgment", "explanation"]
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e.  llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_iters - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+TODO: figure out how these should be supplied to the blocks, and whether any YAML templating based on these are necessary.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+TODO: figure out how to include the appropriate prompt in the block config.
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+TODO: figure out how we want to handle these.
+
+### Model Serving
+
+Custom pipelines may have more unique model serving requirements. Instead of serving just one model, we may need to launch the model server with a model and an additional model with adapter.
+
+How should this be expressed in a flow configuration? Can it be inferred, or are explicit options needed?
```

Comment:
I think there's two parts to this.

Right now the sdg code takes two parameters for each block:

- the openai client instance
- a model ID

I believe these are sufficient. `vllm`, for example, can host both a model and a model+adapter under two different model IDs.

There is another question, maybe separate from this doc, of how the CLI figures out what it's supposed to launch using the internal model serve APIs to do this. I don't know how to handle this yet.

One idea ... instead of a config file loaded by the sdg library, we could make this config file CLI-specific. The config file could have a section that describes the model serving requirements:

```yaml
serve:
- model_id: foo
   model_path: mixtral
- model_id: bar
   model_path: path/to/mixtral
   adapater_path: path/to/adapter

block_configs:
...
```

In this case, we'd probably do all of the config file loading in the `instructlab` repo. Instead of `Flow.from_file()` it could be, `Flow.from_dict()` or something like that.

Reply:
> One idea ... instead of a config file loaded by the sdg library, we could make this config file CLI-specific.

In #86, I converted the default flows to YAML ... and it works well I think. And since it works well, it makes sense to me to not have 2 different ways of defining flows, which helps enforce a level of parity between built-in and custom flow capabilities.

Reply:
Resolving with:

> Right now the sdg code takes two parameters for each block - the OpenAI client instance, and the model ID. This should be sufficient. vLLM, for example, can host both a model and a model+adapter under two di\
fferent model IDs.
>
> In future, it may make sense to	define some model serving config (e.g. which adapter to use for SDG) as	part of	a flow definition. We will address this	in a future if enhancement, if needed.

Reply:
Standardizing on the yaml format everywhere sounds good to me